### ISSUE TITLE
VLAN filtering bridge attach fails when vlan_default_pvid=0

### ISSUE THREAD

**Issue by vuori** [lxc/lxd#10926]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04
 * The output of "lxc info" or if that fails:
 ```
config:
  core.https_address: <removed>
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- snapshot_schedule_aliases
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
- gpu_mdev
- resources_pci_iommu
- resources_network_usb
- resources_disk_address
- network_physical_ovn_ingress_mode
- network_ovn_dhcp
- network_physical_routes_anycast
- projects_limits_instances
- network_state_vlan
- instance_nic_bridged_port_isolation
- instance_bulk_state_change
- network_gvrp
- instance_pool_move
- gpu_sriov
- pci_device_type
- storage_volume_state
- network_acl
- migration_stateful
- disk_state_quota
- storage_ceph_features
- projects_compression
- projects_images_remote_cache_expiry
- certificate_project
- network_ovn_acl
- projects_images_auto_update
- projects_restricted_cluster_target
- images_default_architecture
- network_ovn_acl_defaults
- gpu_mig
- project_usage
- network_bridge_acl
- warnings
- projects_restricted_backups_and_snapshots
- clustering_join_token
- clustering_description
- server_trusted_proxy
- clustering_update_cert
- storage_api_project
- server_instance_driver_operational
- server_supported_storage_drivers
- event_lifecycle_requestor_address
- resources_gpu_usb
- clustering_evacuation
- network_ovn_nat_address
- network_bgp
- network_forward
- custom_volume_refresh
- network_counters_errors_dropped
- metrics
- image_source_project
- clustering_config
- network_peer
- linux_sysctl
- network_dns
- ovn_nic_acceleration
- certificate_self_renewal
- instance_project_move
- storage_volume_project_move
- cloud_init
- network_dns_nat
- database_leader
- instance_all_projects
- clustering_groups
- ceph_rbd_du
- instance_get_full
- qemu_metrics
- gpu_mig_uuid
- event_project
- clustering_evacuation_live
- instance_allow_inconsistent_copy
- network_state_ovn
- storage_volume_api_filtering
- image_restrictions
- storage_zfs_export
- network_dns_records
- storage_zfs_reserve_space
- network_acl_log
- storage_zfs_blocksize
- metrics_cpu_seconds
- instance_snapshot_never
- certificate_token
- instance_nic_routed_neighbor_probe
- event_hub
- agent_nic_config
- projects_restricted_intercept
- metrics_authentication
- images_target_project
- cluster_migration_inconsistent_copy
- cluster_ovn_chassis
- container_syscall_intercept_sched_setscheduler
- storage_lvm_thinpool_metadata_size
- storage_volume_state_total
- instance_file_head
- resources_pci_vpd
- qemu_raw_conf
- storage_cephfs_fscache
- vsock_api
- storage_volumes_all_projects
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: <removed>
  architectures:
  - x86_64
  - i686
  certificate: <removed>
  driver: lxc | qemu
  driver_version: 4.0.12 | 4.2.1
  firewall: xtables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    idmapped_mounts: "true"
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "true"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.13.0-48-generic
  lxc_features:
    cgroup2: "true"
    core_scheduling: "true"
    devpts_fd: "true"
    idmapped_mounts_v2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Ubuntu
  os_version: "20.04"
  project: default
  server: lxd
  server_clustered: false
  server_event_mode: full-mesh
  server_name: maxivalas
  server_pid: 2459049
  server_version: 5.0.1
  storage: btrfs
  storage_version: 5.4.1
  storage_supported_drivers:
  - name: btrfs
    version: 5.4.1
    remote: false
  - name: dir
    version: "1"
    remote: false
  - name: lvm
    version: 2.03.07(2) (2019-11-30) / 1.02.167 (2019-11-30) / 4.45.0
    remote: false
```

# Issue description

Attempting to start a container with a config such as the following fails:

```
devices:
  eth0:
    name: eth0
    nictype: bridged
    parent: br-main
    vlan: 4
    type: nic
```

This results in `Error: Failed to start device "eth0": Failed to run: bridge vlan del dev vethcd2a320a vid 0: exit status 255 (RTNETLINK answers: Invalid argument)`

Notably, br-main has `vlan_default_pvid` set to 0.

Likely cause is that `nic_bridged.go:setupNativeBridgePortVLANs` attempts to unconditionally call `BridgeVLANDelete` with `defaultPVID`, which won't work when `defaultPVID=0`. The delete call is unnecessary in this case, as zero means that no default PVID should be assigned by the kernel.

# Steps to reproduce

 1. Create a bridge and run `/sbin/ip link set dev $IFACE type bridge vlan_filtering 1 vlan_default_pvid 0` to enable VLAN filtering.
 2. Create a container with `bridged` nic attached to bridge created above with `vlan` set to an integer.
 3. Try to start container, note that it fails with above error.

# Information to attach

 - [X] Any relevant kernel output (`dmesg`)
```
[8476729.228251] br-main: port 7(vethcd2a320a) entered blocking state
[8476729.228257] br-main: port 7(vethcd2a320a) entered disabled state
[8476729.228350] device vethcd2a320a entered promiscuous mode
[8476729.230833] device vethcd2a320a left promiscuous mode
[8476729.230861] br-main: port 7(vethcd2a320a) entered disabled state
```
 - [X] Container log (`lxc info NAME --show-log`)
 ```
Name: abuild
Status: STOPPED
Type: container
Architecture: x86_64
Created: 2020/02/15 22:10 UTC
Last Used: 2022/09/18 13:23 UTC

Log:
```
 - [X] Container configuration (`lxc config show NAME --expanded`)
```
architecture: x86_64
config:
  volatile.base_image: d59f8727a7ec00ffa387b434facd67cc4752b8c07be4ef0d7878196380906bb7
  volatile.eth0.hwaddr: <removed>
  volatile.idmap.base: "0"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":165536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":165536,"Nsid":0,"Maprange":65536}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":165536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":165536,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":165536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":165536,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.power: STOPPED
  volatile.uuid: 7abf80aa-24c3-4eec-a100-d1b555825719
devices:
  eth0:
    name: eth0
    nictype: bridged
    parent: br-main
    type: nic
    vlan: "4"
  root:
    path: /
    pool: n2data
    type: disk
ephemeral: false
profiles:
- default2
stateful: false
description: "
```
 - [X] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log): not available; lxd is started with a custom systemd unit file, nothing relevant appears in the journal.
 - [X] Output of the client with --debug
 ```
DEBUG  [2022-09-18T13:42:26Z] Sending request to LXD                        etag= method=GET url="http://unix.socket/1.0/operations/e67d5213-2bf3-4dbf-bcdf-a0be000b2f0f"
DEBUG  [2022-09-18T13:42:26Z] Got response struct from LXD
DEBUG  [2022-09-18T13:42:26Z]
        {
                "id": "e67d5213-2bf3-4dbf-bcdf-a0be000b2f0f",
                "class": "task",
                "description": "Starting instance",
                "created_at": "2022-09-18T13:42:26.193668745Z",
                "updated_at": "2022-09-18T13:42:26.193668745Z",
                "status": "Running",
                "status_code": 103,
                "resources": {
                        "instances": [
                                "/1.0/instances/abuild"
                        ]
                },
                "metadata": null,
                "may_cancel": false,
                "err": "",
                "location": "none"
        }
Error: Failed to start device "eth0": Failed to run: bridge vlan del dev veth004ab59d vid 0: exit status 255 (RTNETLINK answers: Invalid argument)
```
 - [X] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)
```
location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0
    username: root
  level: debug
  message: Handling API request
timestamp: "2022-09-18T13:47:23.638421216Z"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/instances/abuild
    username: root
  level: debug
  message: Handling API request
timestamp: "2022-09-18T13:47:23.64435095Z"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/events
    username: root
  level: debug
  message: Handling API request
timestamp: "2022-09-18T13:47:23.647148189Z"
type: logging


location: none
metadata:
  context:
    id: 98a9d5b3-c42e-4e47-998c-1717021a9056
    local: /var/lib/lxd/unix.socket
    remote: '@'
  level: debug
  message: Event listener server handler started
timestamp: "2022-09-18T13:47:23.647393803Z"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: PUT
    protocol: unix
    url: /1.0/instances/abuild/state
    username: root
  level: debug
  message: Handling API request
timestamp: "2022-09-18T13:47:23.647836879Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    instanceType: container
    project: default
    stateful: "false"
  level: debug
  message: Start started
timestamp: "2022-09-18T13:47:23.657260332Z"
type: logging


location: none
metadata:
  context:
    class: task
    description: Starting instance
    operation: 661cde75-e59e-4d55-8566-da9f92eff9ac
    project: default
  level: debug
  message: Started operation
timestamp: "2022-09-18T13:47:23.657100939Z"
type: logging


location: none
metadata:
  class: task
  created_at: "2022-09-18T13:47:23.649711145Z"
  description: Starting instance
  err: ""
  id: 661cde75-e59e-4d55-8566-da9f92eff9ac
  location: none
  may_cancel: false
  metadata: null
  resources:
    instances:
    - /1.0/instances/abuild
  status: Pending
  status_code: 105
  updated_at: "2022-09-18T13:47:23.649711145Z"
project: default
timestamp: "2022-09-18T13:47:23.657071559Z"
type: operation


location: none
metadata:
  context:
    class: task
    description: Starting instance
    operation: 661cde75-e59e-4d55-8566-da9f92eff9ac
    project: default
  level: debug
  message: New operation
timestamp: "2022-09-18T13:47:23.656985928Z"
type: logging


location: none
metadata:
  class: task
  created_at: "2022-09-18T13:47:23.649711145Z"
  description: Starting instance
  err: ""
  id: 661cde75-e59e-4d55-8566-da9f92eff9ac
  location: none
  may_cancel: false
  metadata: null
  resources:
    instances:
    - /1.0/instances/abuild
  status: Running
  status_code: 103
  updated_at: "2022-09-18T13:47:23.649711145Z"
project: default
timestamp: "2022-09-18T13:47:23.65713835Z"
type: operation


location: none
metadata:
  context:
    action: start
    instance: abuild
    project: default
    reusable: "false"
  level: debug
  message: Instance operation lock created
timestamp: "2022-09-18T13:47:23.657549586Z"
type: logging


location: none
metadata:
  context:
    ip: '@'
    method: GET
    protocol: unix
    url: /1.0/operations/661cde75-e59e-4d55-8566-da9f92eff9ac
    username: root
  level: debug
  message: Handling API request
timestamp: "2022-09-18T13:47:23.658382798Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    project: default
  level: debug
  message: MountInstance started
timestamp: "2022-09-18T13:47:23.659522474Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    project: default
  level: debug
  message: MountInstance finished
timestamp: "2022-09-18T13:47:23.660973524Z"
type: logging


location: none
metadata:
  context:
    err: Network not found
  level: debug
  message: Database error
timestamp: "2022-09-18T13:47:23.661694474Z"
type: logging


location: none
metadata:
  context:
    device: eth0
    instance: abuild
    instanceType: container
    project: default
    type: nic
  level: debug
  message: Starting device
timestamp: "2022-09-18T13:47:23.665199324Z"
type: logging


location: none
metadata:
  context: {}
  level: debug
  message: 'Scheduler: network: veth890c53c9 has been added: updating network priorities'
timestamp: "2022-09-18T13:47:23.666727815Z"
type: logging


location: none
metadata:
  context: {}
  level: debug
  message: 'Scheduler: network: veth0c61edc0 has been added: updating network priorities'
timestamp: "2022-09-18T13:47:23.669629956Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    project: default
  level: debug
  message: UnmountInstance started
timestamp: "2022-09-18T13:47:23.75636622Z"
type: logging


location: none
metadata:
  context:
    action: start
    err: 'Failed to start device "eth0": Failed to run: bridge vlan del dev veth0c61edc0
      vid 0: exit status 255 (RTNETLINK answers: Invalid argument)'
    instance: abuild
    project: default
    reusable: "false"
  level: debug
  message: Instance operation lock finished
timestamp: "2022-09-18T13:47:23.758157625Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    project: default
  level: debug
  message: UnmountInstance finished
timestamp: "2022-09-18T13:47:23.758082614Z"
type: logging


location: none
metadata:
  context:
    class: task
    description: Starting instance
    err: 'Failed to start device "eth0": Failed to run: bridge vlan del dev veth0c61edc0
      vid 0: exit status 255 (RTNETLINK answers: Invalid argument)'
    operation: 661cde75-e59e-4d55-8566-da9f92eff9ac
    project: default
  level: debug
  message: Failure for operation
timestamp: "2022-09-18T13:47:23.758302757Z"
type: logging


location: none
metadata:
  context:
    instance: abuild
    instanceType: container
    project: default
    stateful: "false"
  level: debug
  message: Start finished
timestamp: "2022-09-18T13:47:23.758230666Z"
type: logging


location: none
metadata:
  class: task
  created_at: "2022-09-18T13:47:23.649711145Z"
  description: Starting instance
  err: 'Failed to start device "eth0": Failed to run: bridge vlan del dev veth0c61edc0
    vid 0: exit status 255 (RTNETLINK answers: Invalid argument)'
  id: 661cde75-e59e-4d55-8566-da9f92eff9ac
  location: none
  may_cancel: false
  metadata: null
  resources:
    instances:
    - /1.0/instances/abuild
  status: Failure
  status_code: 400
  updated_at: "2022-09-18T13:47:23.649711145Z"
project: default
timestamp: "2022-09-18T13:47:23.758398969Z"
type: operation


location: none
metadata:
  context:
    listener: 98a9d5b3-c42e-4e47-998c-1717021a9056
    local: /var/lib/lxd/unix.socket
    remote: '@'
  level: debug
  message: Event listener server handler stopped
timestamp: "2022-09-18T13:47:23.760497168Z"
type: logging

```

**Comment by vuori** [comment]
Thanks for the fix, works fine cherry-picked to 5.0.1. Pull-up to stable-5.0 would be great.

**Comment by tomponline** [comment]
Yes it should be backported to next 5.0 LTS release.
